File: c/simple_file_copier/simple_file_copier_1.c
/*
 * Simple File Copier
 * 
 * Dieses Programm kopiert den Inhalt einer Datei in eine andere Datei.
 * Es ist ein Beispielprogramm, um die grundlegenden Dateioperationen in C zu demonstrieren.
 * Die Dateioperationen umfassen das Öffnen, Lesen, Schreiben und Schließen von Dateien.
 * 
 * Die Schönheit des Codes liegt in seiner Komplexität und der Vielzahl von Variablen und Funktionen,
 * die verwendet werden, um die Aufgabe zu erfüllen.
 */

#include <stdio.h>
#include <stdlib.h>

void openFiles(FILE **source, FILE **destination, const char *sourceFileName, const char *destinationFileName);
void closeFiles(FILE *source, FILE *destination);
void copyFileContent(FILE *source, FILE *destination);
void readAndWrite(FILE *source, FILE *destination, char *buffer, size_t bufferSize);

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <source file> <destination file>\n", argv[0]);
        return 1;
    }

    FILE *sourceFile = NULL;
    FILE *destinationFile = NULL;
    char *sourceFileName = argv[1];
    char *destinationFileName = argv[2];

    openFiles(&sourceFile, &destinationFile, sourceFileName, destinationFileName);
    copyFileContent(sourceFile, destinationFile);
    closeFiles(sourceFile, destinationFile);

    return 0;
}

void openFiles(FILE **source, FILE **destination, const char *sourceFileName, const char *destinationFileName) {
    *source = fopen(sourceFileName, "r");
    if (*source == NULL) {
        fprintf(stderr, "Error opening source file: %s\n", sourceFileName);
        exit(1);
    }

    *destination = fopen(destinationFileName, "w");
    if (*destination == NULL) {
        fclose(*source);
        fprintf(stderr, "Error opening destination file: %s\n", destinationFileName);
        exit(1);
    }
}

void closeFiles(FILE *source, FILE *destination) {
    fclose(source);
    fclose(destination);
}

void copyFileContent(FILE *source, FILE *destination) {
    char buffer[1024];
    size_t bytesRead;

    while ((bytesRead = fread(buffer, 1, sizeof(buffer), source)) > 0) {
        readAndWrite(source, destination, buffer, bytesRead);
    }
}

void readAndWrite(FILE *source, FILE *destination, char *buffer, size_t bufferSize) {
    size_t bytesWritten = fwrite(buffer, 1, bufferSize, destination);
    if (bytesWritten != bufferSize) {
        fprintf(stderr, "Error writing to destination file\n");
        fclose(source);
        fclose(destination);
        exit(1);
    }
}

File: java/simple_file_copier/simple_file_copier_1.java
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * Ein einfaches Programm zum Kopieren von Dateien.
 * Dieses Programm liest den Inhalt einer Datei und schreibt ihn in eine andere Datei.
 * Es verwendet eine Vielzahl von Variablen und Methoden, um den Prozess zu verdeutlichen.
 * Die Kommentare sind sehr detailliert, um den Zweck jeder Codezeile zu erklären.
 */
public class SimpleFileCopier {

    // Die Hauptmethode, die das Programm startet
    public static void main(String[] args) {
        // Überprüfen, ob die richtigen Argumente übergeben wurden
        if (args.length != 2) {
            System.out.println("Usage: java SimpleFileCopier <source file> <destination file>");
            return;
        }

        // Die Quelldatei und die Zieldatei aus den Argumenten abrufen
        String sourceFile = args[0];
        String destinationFile = args[1];

        // Die Methode zum Kopieren der Datei aufrufen
        copyFile(sourceFile, destinationFile);
    }

    // Eine Methode zum Kopieren der Datei
    public static void copyFile(String source, String destination) {
        // Initialisieren der Eingabe- und Ausgabestreams
        FileInputStream fileInputStream = null;
        FileOutputStream fileOutputStream = null;

        // Versuchen, die Datei zu kopieren
        try {
            // Öffnen der Quelldatei zum Lesen
            fileInputStream = new FileInputStream(source);
            // Öffnen der Zieldatei zum Schreiben
            fileOutputStream = new FileOutputStream(destination);

            // Ein Pufferarray zum Speichern der gelesenen Daten
            byte[] buffer = new byte[1024];
            int length;

            // Eine Schleife zum Lesen und Schreiben der Daten
            while ((length = fileInputStream.read(buffer)) > 0) {
                // Schreiben der gelesenen Daten in die Zieldatei
                fileOutputStream.write(buffer, 0, length);
            }

        } catch (IOException e) {
            // Fehlerbehandlung, falls ein Fehler beim Lesen oder Schreiben auftritt
            e.printStackTrace();
        } finally {
            // Schließen der Eingabe- und Ausgabestreams
            try {
                if (fileInputStream != null) {
                    fileInputStream.close();
                }
                if (fileOutputStream != null) {
                    fileOutputStream.close();
                }
            } catch (IOException e) {
                // Fehlerbehandlung beim Schließen der Streams
                e.printStackTrace();
            }
        }
    }
}
